src/java.base/share/classes/sun/security/provider/NamedSignature.java

```diff
@@ -0,0 +1,222 @@
+import java.security.ProviderException;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.SignatureException;
+import java.security.SignatureSpi;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Objects;
+
+/// A base class for all `Signature` implementations that can be
+/// configured with a named parameter set. See [NamedKeyPairGenerator]
+/// for more details.
+///
+/// This class does not work with preHash signatures.
+public abstract class NamedSignature extends SignatureSpi {
+
+    private final String fname; // family name
+    private final String[] pnames; // allowed parameter set name (at least one)
+
+    private final ByteArrayOutputStream bout = new ByteArrayOutputStream();
+
+    // init with...
+    private String name = null;
+    private byte[] secKey = null;
+    private byte[] pubKey = null;
+
+    private Object sk2 = null;
+    private Object pk2 = null;
+
+    /// Creates a new `NamedSignature` object.
+    ///
+    /// @param fname the family name
+    /// @param pnames the standard parameter set names, at least one is needed.
+    protected NamedSignature(String fname, String... pnames) {
+        if (fname == null) {
+            throw new AssertionError("fname cannot be null");
+        }
+        if (pnames == null || pnames.length == 0) {
+            throw new AssertionError("pnames cannot be null or empty");
+        }
+        this.fname = fname;
+        this.pnames = pnames;
+    }
+
+    @Override
+    protected void engineInitVerify(PublicKey publicKey) throws InvalidKeyException {
+        // translate also check the key
+        var nk = (NamedX509Key) new NamedKeyFactory(fname, pnames)
+                .engineTranslateKey(publicKey);
+        name = nk.getParams().getName();
+        pubKey = nk.getRawBytes();
+        pk2 = implCheckPublicKey(name, pubKey);
+        secKey = null;
+        bout.reset();
+    }
+
+    @Override
+    protected void engineInitSign(PrivateKey privateKey) throws InvalidKeyException {
+        // translate also check the key
+        var nk = (NamedPKCS8Key) new NamedKeyFactory(fname, pnames)
+                .engineTranslateKey(privateKey);
+        name = nk.getParams().getName();
+        secKey = nk.getRawBytes();
+        sk2 = implCheckPrivateKey(name, secKey);
+        pubKey = null;
+        bout.reset();
+    }
+
+    @Override
+    protected void engineUpdate(byte b) throws SignatureException {
+        bout.write(b);
+    }
+
+    @Override
+    protected void engineUpdate(byte[] b, int off, int len) throws SignatureException {
+        bout.write(b, off, len);
+    }
+
+    @Override
+    protected byte[] engineSign() throws SignatureException {
+        if (secKey != null) {
+            var msg = bout.toByteArray();
+            bout.reset();
+            return implSign(name, secKey, sk2, msg, appRandom);
+        } else {
+            throw new IllegalStateException("No private key");
```

Comment:
`Signature.engineSign` is not specified to throw `IllegalStateException`, even though it seems like the more correct exception to throw. So this should probably be a `SignatureException`.

Same comment for `engineVerify`.

Reply:
In fact, this should never happen since `engineVerify` should only be called when there is a public key (`Signature::verify` checks if `state == VERIFY`). But I'll use `SignatureException`.